### PR TITLE
Fix for RTSP protocol encoding and ecoding

### PR DESCRIPTION
**Pull request by sverkera**
The RTSP protocol isn't strict that only client issues requests and server gives responses, like http protocol. Specifically the server side of a connection may isssue an ANNOUNCE request to the client, which cased the RTSP decoder to go out of sync and fail to decode any more request/responses from the other side.

To solve that issue I've made RtspEncoder and RtspDecoder generic so that they can handle both requests and responses. For compatibility the previous classes RtspRequestEncoder and RtspResponseEncoder now just extends RtspEncoder but are deprecated, likewise for RtspRequestDecoder and RtspResponseDecoder which now extends RtspDecoder.

In RtspDecoderTest there is that specific use case which previously were failing where first an ANNOUNCE request was received and then any other response which made the decoder go out of sync but is now working fine.

### CODE REVIEW

**Comment by sverkera** [comment]
Combine request and response encoder to one class, and vice versa for decoder.

Motivation:
The RTSP protocol isn't strict that only client issues requests and server gives responses, like http protocol. Specifically the server side of a connection may isssue an ANNOUNCE request to the client, which cased the RTSP decoder to go out of sync and fail to decode any more request/responses from the other side.

To solve that issue I've made RtspEncoder and RtspDecoder generic so that they can handle both requests and responses. For compatibility the previous classes RtspRequestEncoder and RtspResponseEncoder now just extends RtspEncoder but are deprecated, likewise for RtspRequestDecoder and RtspResponseDecoder which now extends RtspDecoder.

In RtspDecoderTest there is that specific use case which previously were failing where first an ANNOUNCE request was received and then any other response which made the decoder go out of sync but is now working fine.

Modifications:
RtspObjectEncoder refactored to RtspEncoder which is now the main encoder class.
RtspRequestEncoder and RtspResponseEncoder are kept for backward compatibility but just extends RtspEncoder and are deprecated.

Likewise RtspObjectDecoder was refactored to RtspDecoder which is now the main decoder class. RtspReqestDecoder and RtspResponseDecoder are kept  for backward compatibility but just extends RtspDecoder and are deprecated.

Result:
The RTSP protocol encoder and decoder is now able to send and receive both requests and response in the same channel.

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderNames.java [hunk 1]
```diff
@@ -22,186 +22,234 @@
 /**
  * Standard RTSP header names.
  * <p>
- * These are all defined as lowercase to support HTTP/2 requirements while also not
- * violating RTSP/1.x requirements.  New header names should always be lowercase.
+ * These are all defined as lowercase to support HTTP/2 requirements while also
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderNames.java]: Please revert the changes in this class as these are not really are needed for the functionality

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import io.netty.handler.codec.http.HttpObjectEncoder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.util.CharsetUtil;
+
+/**
+ * Encodes an RTSP message represented in {@link FullHttpMessage} into
+ * a {@link ByteBuf}.
+ */
+@Sharable
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: HttpObjectEncoder has a state so this can not be `@Sharable`

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpObjectEncoder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.util.CharsetUtil;
+
+/**
+ * Encodes an RTSP message represented in {@link FullHttpMessage} into
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: I think this javadoc is not correct as it not only handle `FullHttpMessage` but also `HttpMessage`, `HttpContent` etc.

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            buf.writeByte(SP);
+            buf.writeBytes(String.valueOf(response.status().code())
+                                 .getBytes(CharsetUtil.US_ASCII));
+            buf.writeByte(SP);
+            HttpHeaders.encodeAscii(String.valueOf(response.status()
+                                                           .reasonPhrase()),
+                                    buf);
+            buf.writeBytes(CRLF);
+        } else {
+            throw new Exception("Unsupported type "
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: throw UnsupportedMessageTypeException

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * True if the message to decode is a request.
+     * False if the message to decode is a response.
+     */
+    private boolean isDecodingRequest;
+
+    /**
+     * Regex used on first line in message to detect if it is a response.
+     */
+    private String versionRegex = "RTSP/\\d\\.\\d";
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: Use a precompile static Pattern for performance reasons.

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspMethods.java [hunk 6]
```diff
@@ -26,11 +26,12 @@
 public final class RtspMethods {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspMethods.java]: please revert changes in this method as these are not needed for the functionality

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspResponseStatuses.java [hunk 7]
```diff
@@ -23,270 +23,370 @@
 public final class RtspResponseStatuses {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspResponseStatuses.java]: Please revert changes in this class as these are not needed for the functionality

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspVersions.java [hunk 8]
```diff
@@ -23,29 +23,35 @@
 public final class RtspVersions {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspVersions.java]: Please revert the changes as these are not needed for the functionality

File: codec-http/src/test/java/io/netty/handler/codec/rtsp/RtspDecoderTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2013 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/rtsp/RtspDecoderTest.java]: 2015

File: codec-http/src/test/java/io/netty/handler/codec/rtsp/RtspEncoderTest.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2013 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/rtsp/RtspEncoderTest.java]: 2015

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: 2015

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: 2015

**Comment by normanmaurer** [comment]
@sverkera thanks for your work! Please address my comments and ping me once done :)

**Comment by sverkera** [comment]
@normanmaurer ping ;). I believe I've covered all your comments now, pls let me know if I missed anything.

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        // Unlike HTTP, RTSP always assumes zero-length body if Content-Length
+        // header is absent.
+        boolean empty = super.isContentAlwaysEmpty(msg);
+        if (empty) {
+            return true;
+        }
+        if (!msg.headers().contains(RtspHeaderNames.CONTENT_LENGTH)) {
+            return true;
+        }
+        return empty;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: I think this can be simplified as:

```java
    return super.isContentAlwaysEmpty(msg) || !msg.headers().contains(RtspHeaderNames.CONTENT_LENGTH);
```

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderValues.java [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014 The Netty Project
+ * Copyright 2015 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderValues.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderNames.java [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014 The Netty Project
+ * Copyright 2015 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaderNames.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspMethods.java [hunk 16]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012 The Netty Project
+ * Copyright 2015 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspMethods.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspVersions.java [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012 The Netty Project
+ * Copyright 2015 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspVersions.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspResponseStatuses.java [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012 The Netty Project
+ * Copyright 2015 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspResponseStatuses.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/package-info.java [hunk 19]
```diff
@@ -15,7 +15,8 @@
  */
 
 /**
- * An <a href="http://en.wikipedia.org/wiki/Real_Time_Streaming_Protocol">RTSP</a>
+ * An
+ * <a href="http://en.wikipedia.org/wiki/Real_Time_Streaming_Protocol">RTSP</a>
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/package-info.java]: revert

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspObjectDecoder.java [hunk 20]
```diff
@@ -1,87 +0,0 @@
...
- *     header exceeds this value, a {@link TooLongFrameException} will be raised.</td>
- * </tr>
- * <tr>
- * <td>{@code maxContentLength}</td>
- * <td>The maximum length of the content.  If the content length exceeds this
- *     value, a {@link TooLongFrameException} will be raised.</td>
- * </tr>
- * </table>
- */
-public abstract class RtspObjectDecoder extends HttpObjectDecoder {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspObjectDecoder.java]: we will need to keep it and mark as @deprecated

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspObjectEncoder.java [hunk 21]
```diff
@@ -1,41 +0,0 @@
...
-import io.netty.handler.codec.http.FullHttpMessage;
-import io.netty.handler.codec.http.HttpMessage;
-import io.netty.handler.codec.http.HttpObjectEncoder;
-
-/**
- * Encodes an RTSP message represented in {@link FullHttpMessage} into
- * a {@link ByteBuf}.
- */
-@Sharable
-public abstract class RtspObjectEncoder<H extends HttpMessage> extends HttpObjectEncoder<H> {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspObjectEncoder.java]: we will need to keep it and mark as @deprecated

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+    /**
+     * True if the message to decode is a request.
+     * False if the message to decode is a response.
+     */
+    private boolean isDecodingRequest;
+
+    /**
+     * Regex used on first line in message to detect if it is a response.
+     */
+    private static Pattern versionPattern = Pattern.compile("RTSP/\\d\\.\\d");
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: static final

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+     * Constant for CRLF.
+     */
+    private static final byte[] CRLF = {CR, LF};
+
+    @Override
+    public boolean acceptOutboundMessage(final Object msg)
+           throws Exception {
+        return super.acceptOutboundMessage(msg)
+                &&
+                ((msg instanceof HttpRequest) || (msg instanceof HttpResponse));
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: merge the two lines

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    protected void encodeInitialLine(final ByteBuf buf,
+                                           final HttpMessage message)
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: merge the two lines

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+            buf.writeBytes(String.valueOf(response.status().code())
+                                 .getBytes(CharsetUtil.US_ASCII));
+            buf.writeByte(SP);
+            HttpHeaders.encodeAscii(String.valueOf(response.status()
+                                                           .reasonPhrase()),
+                                    buf);
+            buf.writeBytes(CRLF);
+        } else {
+            throw new UnsupportedMessageTypeException("Unsupported type "
+                                + message.getClass().getName());
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspEncoder.java]: StringUtil.simpleName(....)

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaders.java [hunk 26]
```diff
@@ -24,6 +24,7 @@
  * Standard RTSP header names and values.
  */
 @Deprecated
+@SuppressWarnings("deprecation")
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaders.java]: revert ?

Comment (sverkera) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspHeaders.java]: Could be, I usually supress warnings of use of deprecated other classes in a deprecated class because typically that otherwise results in a lot of warnings that are obsolete and which may hide warnings that are relevant. Your call, I'll revert it if you prefer so.

**Comment by normanmaurer** [comment]
@sverkera please squash all your commits and sign our ICLA:
http://netty.io/s/icla

**Comment by sverkera** [comment]
@normanmaurer I hope I got the squash right now. ICLA is signed.

**Comment by normanmaurer** [comment]
@sverkera thanks... reviewing :)

File: codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+        // Unlike HTTP, RTSP always assumes zero-length body if Content-Length
+        // header is absent.
+        return super.isContentAlwaysEmpty(msg) || !msg.headers().contains(RtspHeaderNames.CONTENT_LENGTH);
+    }
+
+    @Override
+    protected HttpMessage createInvalidMessage() {
+        if (isDecodingRequest) {
+            return new DefaultFullHttpRequest(RtspVersions.RTSP_1_0,
+                       RtspMethods.OPTIONS, "/bad-request", validateHeaders);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: is OPTIONS right here ?

Comment (sverkera) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: If you look below at RtspRequestDecoder this is consistent with how it was handled before. In reality this is a no-op, there has to be a valid method and OPTIONS is harmless in that aspect but it is the path "/bad-request" that is significant. The other alternative would be to add an UNKNOWN method to RtspMethods, which might be better design wise, but that would be specific for this implementation.

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/rtsp/RtspDecoder.java]: I see.. Sorry for the delay.

**Comment by normanmaurer** [comment]
@sverkera can you rebase one last time on 4.1 and push ?

**Comment by normanmaurer** [comment]
@sverkera thanks again for your patience...

Cherry-picked into 4.0 (3b2648c3d1c8d0692e68e325cfa90580997344bf), 4.1 (e121c68e0fb633b1fceca24ad0f5b04d14879990) and master (7888ae8d296379d7a095d84395efc1c16d618889)

**Comment by sverkera** [comment]
Thank you
